rust/src/derivation.rs

```diff
@@ -0,0 +1,539 @@
+pub unsafe extern "C" fn zcashlc_free_typecodes(data: *mut u32, len: usize) {
+    free_ptr_from_vec(data, len);
+}
+
+/// Returns true when the provided key decodes to a valid Sapling extended spending key for the
+/// specified network, false in any other case.
+///
+/// # Safety
+///
+/// - `extsk` must be non-null and must point to a null-terminated UTF-8 string.
+/// - The memory referenced by `extsk` must not be mutated for the duration of the function call.
+#[no_mangle]
+pub unsafe extern "C" fn zcashlc_is_valid_sapling_extended_spending_key(
+    extsk: *const c_char,
+    network_id: u32,
+) -> bool {
+    let res = catch_panic(|| {
+        let network = parse_network(network_id)?;
+        let extsk = unsafe { CStr::from_ptr(extsk).to_str()? };
+
+        Ok(
+            decode_extended_spending_key(network.hrp_sapling_extended_spending_key(), extsk)
+                .is_ok(),
+        )
+    });
+    unwrap_exc_or(res, false)
+}
+
+/// Returns true when the provided key decodes to a valid Sapling extended full viewing key for the
+/// specified network, false in any other case.
+///
+/// # Safety
+///
+/// - `key` must be non-null and must point to a null-terminated UTF-8 string.
+/// - The memory referenced by `key` must not be mutated for the duration of the function call.
+#[no_mangle]
+pub unsafe extern "C" fn zcashlc_is_valid_viewing_key(key: *const c_char, network_id: u32) -> bool {
+    let res =
+        catch_panic(|| {
+            let network = parse_network(network_id)?;
+            let vkstr = unsafe { CStr::from_ptr(key).to_str()? };
+
+            Ok(decode_extended_full_viewing_key(
+                network.hrp_sapling_extended_full_viewing_key(),
+                vkstr,
+            )
+            .is_ok())
+        });
+    unwrap_exc_or(res, false)
+}
+
+/// Returns true when the provided key decodes to a valid unified full viewing key for the
+/// specified network, false in any other case.
+///
+/// # Safety
+///
+/// - `ufvk` must be non-null and must point to a null-terminated UTF-8 string.
+/// - The memory referenced by `ufvk` must not be mutated for the duration of the
+///   function call.
+#[no_mangle]
+pub unsafe extern "C" fn zcashlc_is_valid_unified_full_viewing_key(
+    ufvk: *const c_char,
+    network_id: u32,
+) -> bool {
+    let res = catch_panic(|| {
+        let network = parse_network(network_id)?;
+        let ufvkstr = unsafe { CStr::from_ptr(ufvk).to_str()? };
+
+        Ok(UnifiedFullViewingKey::decode(&network, ufvkstr).is_ok())
+    });
+    unwrap_exc_or(res, false)
+}
+
+/// Derives and returns a unified spending key from the given seed for the given account ID.
+///
+/// Returns the binary encoding of the spending key. The caller should manage the memory of (and
+/// store, if necessary) the returned spending key in a secure fashion.
+///
+/// # Safety
+///
+/// - `seed` must be non-null and valid for reads for `seed_len` bytes, and it must have an
+///   alignment of `1`.
```

Comment:
Having an alignment of 1 imposes no requirement (and you would not expect it to have any other alignment, having type `*const u8`).